lib/src/rules/prefer_is_empty.dart

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.prefer_is_empty;
+
+import 'package:analyzer/context/declared_variables.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:analyzer/error/listener.dart';
+import 'package:analyzer/src/dart/constant/evaluation.dart';
+import 'package:analyzer/src/dart/constant/value.dart';
+import 'package:analyzer/src/error/codes.dart';
+import 'package:analyzer/src/generated/engine.dart';
+import 'package:analyzer/src/generated/resolver.dart';
+import 'package:analyzer/src/lint/linter.dart';
+
+const desc = 'Use isNotEmpty for Iterables and Maps.';
+
+const useIsNotEmpty = 'Use isNotEmpty instead of length';
+const useIsEmpty = 'Use isEmpty instead of length';
+const alwaysFalse = 'Always false because length is always greater or equal 0.';
+const alwaysTrue = 'Always true because length is always greater or equal 0.';
+
+const details = '''
+**DO NOT** use `.length` to see if a collection is empty.
+
+The `Iterable` contract does not require that a collection know its length or be
+able to provide it in constant time. Calling `.length` just to see if the
+collection contains anything can be painfully slow.
+
+Instead, there are faster and more readable getters: `.isEmpty` and
+`.isNotEmpty`. Use the one that doesn’t require you to negate the result.
+
+**GOOD:**
+```
+if (lunchBox.isEmpty) return 'so hungry...';
+if (words.isNotEmpty) return words.join(' ');
+```
+
+**BAD:**
+```
+if (lunchBox.length == 0) return 'so hungry...';
+if (words.length != 0) return words.join(' ');
+```
+''';
+
+class _LintCode extends LintCode {
```

Comment:
This appears in multiple rules. As a separate PR (by you or someone else) we should move this to a utility library so that we don't have to keep replicating the code. Not something that needs to be addressed now.